lib/controlled-component.js

```diff
@@ -17,24 +17,13 @@ export default class ControlledComponent extends Component {
   detachedCallback() {
     super.detachedCallback();
     this._unsubscribeUpdates();
-    this._lastState = null; // No dangling references when cleaning up
   }
 
   attributeChangedCallback() {
     // Do nothing, component should explicitly pass this to controller for an update
     // Super class calls this.update() which will throw an error
   }
 
-  // Don't render if state hasn't changed since controlled components return immutable state
-  shouldUpdate(newState) {
-    if (this._lastState !== newState) {
-      this._lastState = newState;
-      return true;
-    }
```

Comment:
What would be cool is if StateControllers could be responsible for a subtree of state instead of the whole thing, and components could do something like
```
shouldUpdate(newState) {
  this.controller.shouldUpdate(newState);
}
```
where the controller makes a check like you have above and only updates if the relevant subtree has changed. This would let us not re-render every single component on every single update - similar to the "pure component" style of building a react app, something I'm a big fan of and have wanted to emulate in panel. imo this is one of the big benefits of keeping all state changes immutable since you can then do this with constant-time state obj comparisons instead of deeply comparing state